mesonbuild/compilers/c.py

```diff
@@ -0,0 +1,996 @@
+        result = []
+        for i in args:
+            # -mms-bitfields is specific to MinGW-GCC
+            # -pthread is only valid for GCC
+            if i in ('-mms-bitfields', '-pthread'):
+                continue
+            if i.startswith('-L'):
+                i = '/LIBPATH:' + i[2:]
+            # Translate GNU-style -lfoo library name to the import library
+            elif i.startswith('-l'):
+                name = i[2:]
+                if name in ('m', 'c', 'pthread'):
+                    # With MSVC, these are provided by the C runtime which is
+                    # linked in by default
+                    continue
+                else:
+                    i = name + '.lib'
+            # -pthread in link flags is only used on Linux
+            elif i == '-pthread':
+                continue
+            result.append(i)
+        return result
+
+    def get_werror_args(self):
+        return ['/WX']
+
+    def get_include_args(self, path, is_system):
+        if path == '':
+            path = '.'
+        # msvc does not have a concept of system header dirs.
+        return ['-I' + path]
+
+    # Visual Studio is special. It ignores some arguments it does not
+    # understand and you can't tell it to error out on those.
+    # http://stackoverflow.com/questions/15259720/how-can-i-make-the-microsoft-c-compiler-treat-unknown-flags-as-errors-rather-t
+    def has_multi_arguments(self, args, env):
+        warning_text = '9002'
+        code = 'int i;\n'
+        (fd, srcname) = tempfile.mkstemp(suffix='.' + self.default_suffix)
+        os.close(fd)
+        with open(srcname, 'w') as ofile:
+            ofile.write(code)
+        # Read c_args/cpp_args/etc from the cross-info file (if needed)
+        extra_args = self.get_cross_extra_flags(env, link=False)
+        extra_args += self.get_compile_only_args()
+        commands = self.exelist + args + extra_args + [srcname]
+        mlog.debug('Running VS compile:')
+        mlog.debug('Command line: ', ' '.join(commands))
+        mlog.debug('Code:\n', code)
+        p, stdo, stde = Popen_safe(commands, cwd=os.path.split(srcname)[0])
+        if p.returncode != 0:
+            return False
+        return not(warning_text in stde or warning_text in stdo)
+
+    def get_compile_debugfile_args(self, rel_obj, pch=False):
+        pdbarr = rel_obj.split('.')[:-1]
+        pdbarr += ['pdb']
+        args = ['/Fd' + '.'.join(pdbarr)]
+        # When generating a PDB file with PCH, all compile commands write
+        # to the same PDB file. Hence, we need to serialize the PDB
+        # writes using /FS since we do parallel builds. This slows down the
+        # build obviously, which is why we only do this when PCH is on.
+        # This was added in Visual Studio 2013 (MSVC 18.0). Before that it was
+        # always on: https://msdn.microsoft.com/en-us/library/dn502518.aspx
+        if pch and mesonlib.version_compare(self.version, '>=18.0'):
```

Comment:
__[flake8]__
* [F405] 'mesonlib' may be undefined, or defined from star imports: .compilers

<!-- Comment by [SideCI](https://sideci.com) -->